internal/engine/buildcontroller_test.go

```diff
@@ -1047,3 +1048,303 @@ func TestLogsLongResourceName(t *testing.T) {
+	f.assertAllBuildsConsumed()
+}
+
+func (f *testFixture) waitUntilManifestBuilding(name model.ManifestName) {
+	msg := fmt.Sprintf("manifest %q is building", name)
+	f.WaitUntilManifestState(msg, name, func(ms store.ManifestState) bool {
+		return ms.IsBuilding()
+	})
+
+	f.withState(func(st store.EngineState) {
+		_, ok := st.CurrentlyBuilding[name]
+		require.True(f.t, ok, "expected EngineState to reflect that %q is currently building", name)
+	})
+}
+
+func (f *testFixture) waitUntilBuildCountAtLeast(n int) {
+	ctx, cancel := context.WithTimeout(f.ctx, time.Millisecond*200)
+	defer cancel()
+	for {
+		if f.b.buildCount >= n {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			f.T().Fatalf("Timed out waiting for buildCount >= %d", n)
+		case <-time.After(time.Millisecond * 5):
+		}
+	}
+}
+
+func (f *testFixture) waitUntilManifestNotBuilding(name model.ManifestName) {
+	msg := fmt.Sprintf("manifest %q is NOT building", name)
+	f.WaitUntilManifestState(msg, name, func(ms store.ManifestState) bool {
+		return !ms.IsBuilding()
+	})
+
+	f.withState(func(st store.EngineState) {
+		_, ok := st.CurrentlyBuilding[name]
+		require.False(f.t, ok, "expected EngineState to reflect that %q is NOT currently building", name)
+	})
+}
+
+func (f *testFixture) waitUntilNumBuildSlots(expected int) {
+	msg := fmt.Sprintf("%d build slots available", expected)
+	f.WaitUntil(msg, func(st store.EngineState) bool {
+		return expected == st.AvailableBuildSlots()
+	})
+}
+
+func (f *testFixture) editFileAndWaitForManifestBuilding(name model.ManifestName, path string) (buildIndex int) {
+	startBuildCount := f.b.buildCount
+	f.fsWatcher.events <- watch.NewFileEvent(f.JoinPath(path))
+	f.waitUntilManifestBuilding(name)
+	f.waitUntilBuildCountAtLeast(startBuildCount + 1)
+	return startBuildCount + 1
+}
+
+func (f *testFixture) editFileAndAssertManifestNotBuilding(name model.ManifestName, path string) {
+	f.fsWatcher.events <- watch.NewFileEvent(f.JoinPath(path))
+	f.waitUntilManifestNotBuilding(name)
+}
+
+func (f *testFixture) runAndWaitForManifestBuilding(name model.ManifestName, fn func()) (buildIndex int) {
+	startBuildCount := f.b.buildCount
+	fn()
+	f.waitUntilManifestBuilding(name)
+	f.waitUntilBuildCountAtLeast(startBuildCount + 1)
+	return startBuildCount + 1
+}
+
+func (f *testFixture) assertCallIsForManifestAndFiles(call buildAndDeployCall, m model.Manifest, files ...string) {
+	assert.Equal(f.t, m.ImageTargetAt(0).ID(), call.firstImgTarg().ID())
+	assert.Equal(f.t, f.JoinPaths(files), call.oneState().FilesChanged())
+}
+
+func (f *testFixture) completeAndCheckBuildsForManifests(indexes []int, manifests []model.Manifest) {
+	if len(indexes) != len(manifests) {
+		f.t.Fatalf("can only complete + check builds for manifests when passed an equal number of indexes and manifests")
+	}
+
+	for _, i := range indexes {
+		f.b.completeBuild(i)
+	}
+
+	targetsFromCalls := f.imageTargsForNextNCalls(len(manifests))
+	require.Len(f.t, targetsFromCalls, len(manifests))
```

Comment:
this check doesn't seem correct, and only works in particular tests

Reply:
fixed!